allennlp/modules/seq2seq_decoders/stacked_self_attention_decoder_net.py

```diff
@@ -0,0 +1,149 @@
+                                                                                        SublayerConnection,
+                                                                                        PositionalEncoding,
+                                                                                        MultiHeadedAttention)
+from allennlp.modules.seq2seq_decoders.decoder_net import DecoderNet
+from allennlp.modules.layer_norm import LayerNorm
+
+
+@DecoderNet.register("stacked_self_attention")
+class StackedSelfAttentionDecoderNet(DecoderNet):
+    """
+    A Stacked self-attention decoder implementation.
+
+    Parameters
+    ----------
+    decoding_dim : ``int``, required
+        Defines dimensionality of output vectors.
+    target_embedding_dim : ``int``, required
+        Defines dimensionality of input target embeddings.  Since this model takes it's output on a previous step
+        as input of following step, this is also an input dimensionality.
+    feedforward_hidden_dim : ``int``, required.
+        The middle dimension of the FeedForward network. The input and output
+        dimensions are fixed to ensure sizes match up for the self attention layers.
+    num_layers : ``int``, required.
+        The number of stacked self attention -> feedfoward -> layer normalisation blocks.
+    num_attention_heads : ``int``, required.
+        The number of attention heads to use per layer.
+    use_positional_encoding: ``bool``, optional, (default = True)
+        Whether to add sinusoidal frequencies to the input tensor. This is strongly recommended,
+        as without this feature, the self attention layers have no idea of absolute or relative
+        position (as they are just computing pairwise similarity between vectors of elements),
+        which can be important features for many tasks.
+    dropout_prob : ``float``, optional, (default = 0.1)
+        The dropout probability for the feedforward network.
+    residual_dropout_prob : ``float``, optional, (default = 0.2)
+        The dropout probability for the residual connections.
+    attention_dropout_prob : ``float``, optional, (default = 0.1)
+        The dropout probability for the attention distributions in each attention layer.
+    """
+
+    def __init__(self,
+                 decoding_dim: int,
+                 target_embedding_dim: int,
+                 feedforward_hidden_dim: int,
+                 num_layers: int,
+                 num_attention_heads: int,
+                 use_positional_encoding: bool = True,
+                 positional_encoding_max_steps: int = 5000,
+                 dropout_prob: float = 0.1,
+                 residual_dropout_prob: float = 0.2,
+                 attention_dropout_prob: float = 0.1,) -> None:
+
+        super().__init__(decoding_dim=decoding_dim,
+                         target_embedding_dim=target_embedding_dim,
+                         decodes_parallel=True)
+
+        deep_copy = copy.deepcopy
```

Comment:
I think saying `from copy import deepcopy` would be more typical unless I'm missing something.